timdex_dataset_api/record.py

```diff
@@ -16,40 +19,34 @@ class DatasetRecord:
     """
 
     # primary columns
-    timdex_record_id: str
-    source_record: bytes
-    transformed_record: bytes
-
-    # partition columns
-    source: str | None = None
-    run_date: str | datetime.datetime | None = None
-    run_type: str | None = None
-    run_id: str | None = None
-    action: str | None = None
+    timdex_record_id: str = field()
+    source_record: bytes = field()
+    transformed_record: bytes = field()
+    source: str = field()
+    run_date: datetime = field(converter=strict_date_parse)
```

Comment:
It's worth noting that the `__init__` method assumes that users are passing (a) strings (b) strictly formatted as "YYYY-MM-DD" (Python date format code `"%Y-%m-%d"`). Initially, we were allowing users to provide either a `string` _or_ a `datetime.datetime` object but that introduces some complexity...Since we are primarily using the `run_date` field to derive `year, month, day` values, I think what we have now is sufficient.

Reply:
At first I was a bit dissappointed that we can no longer instantiate (or set) `run_date` with a datetime object, given that's effectively our target format.  But, if development of this library has demonstrated anything, it's that it may yet require change as we progress.  Keeping it simple and strict now, feels like a great path; easier to extend functionality later than reign in expansive options/features when we don't even have concrete use cases for.